src/sentry/search/events/builder.py

```diff
@@ -1552,6 +1553,127 @@ def get_snql_query(self) -> None:
         inmplemented see run_query"""
         raise NotImplementedError("get_snql_query cannot be implemented for MetricsQueryBuilder")
 
-    def run_query(self) -> None:  # type: ignore
-        # TODO(wmak): will implement this soon
-        pass
+    def run_query(self, referrer: str, use_cache: bool = False) -> Any:
+        # Need to split orderby between the 3 possible tables
+        query_framework = {
+            "distribution": {
+                "orderby": [],
+                "functions": self.distributions,
+                "entity": "metrics_distributions",
+            },
+            "counter": {
+                "orderby": [],
+                "functions": self.counters,
+                "entity": "metrics_counters",
+            },
+            "set": {
+                "orderby": [],
+                "functions": self.sets,
+                "entity": "metrics_sets",
+            },
+        }
+        general_orderby = []
+        primary = None
+        # if orderby spans more than one table, the query isn't possible with metrics
+        for orderby in self.orderby:
+            if orderby.exp in self.distributions:
+                query_framework["distribution"]["orderby"].append(orderby)
+                if primary not in [None, "distribution"]:
+                    raise IncompatibleMetricsQuery("Can't order across tables")
+                primary = "distribution"
+            elif orderby.exp in self.sets:
+                query_framework["set"]["orderby"].append(orderby)
+                if primary not in [None, "set"]:
+                    raise IncompatibleMetricsQuery("Can't order across tables")
+                primary = "set"
+            elif orderby.exp in self.counters:
+                query_framework["counter"]["orderby"].append(orderby)
+                if primary not in [None, "counter"]:
+                    raise IncompatibleMetricsQuery("Can't order across tables")
+                primary = "counter"
+            else:
+                general_orderby.append(orderby)
+
+        # Pick one arbitrarily, there's no orderby on functions
+        if primary is None:
+            primary = "distribution"
+
+        value_map = defaultdict(dict)
+        groupby_aliases = [
+            groupby.alias
+            if isinstance(groupby, (AliasedExpression, CurriedFunction))
+            else groupby.name
+            for groupby in self.groupby
+        ]
+        groupby_values = []
+        result = {
+            "data": None,
+            "meta": [],
+        }
+        for query_details in [query_framework.pop(primary), *query_framework.values()]:
```

Comment:
This is kind of fine, but it'd probably be clearer to just do a single request with the primary then loop through the rest imo. If that's too much repeated code, then take the block below and turn it into a function.